### DISCUSSION TITLE
Skeleton instead of default loading behaviour for Datatable

### DISCUSSION THREAD

**Discussion by setasan** [primefaces/community#1128] [PrimeReact] [answered]
Is there a way to add a skeleton for each row when using prop loading for the datatable? I only managed to use loadingIcon prop but I'd like to use skeleton instead. I noticed there is a prop called virtualScrollerOptions that have a attribute 'loadingTemplate' which is exacty what I want, but since I am not using scroller it has no use to me.

**Comment by melloware** [comment, marked as answer]
See this workaround: https://stackblitz.com/edit/mxk9cf?file=src%2FApp.tsx

```js
import "primereact/resources/themes/lara-light-indigo/theme.css";
import "primereact/resources/primereact.min.css";
import "primeicons/primeicons.css";

import { DataTable } from "primereact/datatable";
import { Skeleton } from "primereact/skeleton";
import { Column } from "primereact/column";
import { VirtualScrollerLoadingTemplateType } from "primereact/virtualscroller";

type DataType = {
  id: number;
  code: number;
  description: string;
};

const data = () => {
  const preData: DataType[] = [];
  for (let i = 0; i < 1000; i++) {
    preData.push({
      id: i,
      code: i,
      description: "desc " + i
    });
  }
  return preData;
};

const loadingTemplate: VirtualScrollerLoadingTemplateType = (options) => (
  <div
    {...{
      style: {
        height: 30,
        width: "100%",
        overflow: "hidden",
        display: "flex",
        alignItems: "center",
        justifyContent: "flex-start"
      }
    }}
  >
    <Skeleton
      {...{
        style: {
          paddingLeft: 10
        },
        // if column is either checkbox row selector or action buttons,
        // then its sekeleton should be a fixed width of 90%
        // else varies width according to options.even
        width: options.even ? "25%" : "45%",
        height: "60%"
      }}
    />
  </div>
);

export default function App() {
  return (
    <DataTable
      value={data()}
      // virtual scroll
      scrollable
      scrollHeight="400px"
      virtualScrollerOptions={{
        delay: 200,
        itemSize: 30,
        showLoader: true,
        loadingTemplate
      }}
      // resize
      showGridlines={true}
      resizableColumns={true}
    >
      <Column columnKey="id" field="id" header="Id" />
      <Column columnKey="code" field="code" header="Code" />
      <Column columnKey="desc" field="description" header="Description" />
    </DataTable>
  );
}

```

**Comment by modim30** [comment]
I dont think this will work.

**Reply by melloware** [reply]
@absogod did you actually try the StackBlitz?  it works fine: https://stackblitz.com/edit/mxk9cf?file=src%2FApp.tsx

**Reply by modim30** [reply]
I want to implement the skeleton without virtualScroller, as I am using pagination for the table. So I dont think this will work.

**Comment by cacdivic** [comment]
@melloware Hi, I have a question regarding customizing the loading state for DataTable.

Currently, I’m not using virtual scroll, but I still want to show a loading indicator — specifically using Skeleton components while data is being fetched.

At the moment, my workaround is:
	•	I conditionally set value to an array of empty dummy items during loading.
	•	Inside each column’s body, I check the loading state and render <Skeleton /> accordingly.

This works, but I’m wondering —
Is there a more official or recommended way to show a loading UI like Skeletons without virtual scroll?

Thanks a lot!

**Reply by melloware** [reply]
Nope no official way your way sounds as good as any
